---
title: "Creating and Comparing Phylogenetic Trees in R"
author: "Abby, Isabel, Marta, Nirmiti"
output: 
  html_document:
    toc: true
    toc_float: true
    css: bootstrap.css
    number_sections: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ape)
library(phytools)
```

<style>
div.blue pre { background-color:navy; }
div.blue pre.r { background-color:navy; }
</style>


# Phylogentic Tree Comparison

<p class="text-primary">In our module we want to break down phylogentic tress and understand how R coding can help us compare them.</p>

## Packages

>For starters, you will need to install [{ape}](https://www.rdocumentation.org/packages/ape/versions/5.5), [{phytools}](https://www.rdocumentation.org/packages/phytools/versions/0.7-90)

## Objectives

In this module we will teach you how to:  
- Create trees and understand phylogenetic trees  
- Upload and forma data into trees  
- Modify trees using R packages  
- And compare morphological and molecular trees  


## Introduction 

Evolutionary biologists study the ways organisms change over time. This is often done by comparing several species to one another and analyzing the differences between them. These differences can then be used to build a graphical representation of species’ relationships to one another, called a phylogenetic tree. Originally, these trees were constructed based solely on organism morphology using shared derived traits, or __synapomorphies__, often in the form of character matrices. By focusing on these physical traits, biologists could hypothesize different life histories, like species divergence or common descent. However, using morphological data to determine taxonomic relationships is not foolproof. Some shared, analogous traits may be due to convergent evolution. Homoplasy can be difficult to detect through morphological analysis alone, and has historically resulted in phylogenetic trees that have since been disproved through the use of molecular data.
	
Modern biologists can now build phylogenetic trees based on the DNA of the organisms in question. By focusing on a few specific genes, evolutionary relationships can be hypothesized with greater accuracy. Examining the polymorphism at different loci can provide information on how distantly related organisms are to one another, and can reduce errors commonly associated with comparing or identifying cryptic species. Molecular data can help resolve unclear phylogenies that were created before genomic methods were available. Biologists can now revisit unresolved trees built from morphological data and revise them using DNA, allowing us to better hypothesize evolutionary relationships.
	
In order to assess any changes between these hypothesized evolutionary relationships, we can compare the topologies of phylogenetic trees. The topology refers to the branching pattern displayed, which represents the measure of relatedness among taxa. To understand the importance of topology, we must familiarize ourselves with the anatomy of a phylogenetic tree. 

<h4><p class="text-primary">Anatomy of a tree:</p></h4> 

![__Why would we compare tree topologies?__](https://media.springernature.com/full/springer-static/image/art%3A10.1007%2Fs12052-008-0035-x/MediaObjects/12052_2008_35_Fig2_HTML.gif?as=webp)

## Methods

There are several different ways to estimate phylogenies, each with their own strengths, weaknesses, and appropriate situations in which to apply them. While we won’t go over them in detail, familiarize yourselves by looking at <a href="https://fuzzyatelin.github.io/bioanth-stats/module-24/module-24.html)">Module 24</a>  
There are MANY different phylogenetics software that we can use to employ these methods of estimation, but some of the most common are PAUP, BEAST, MrBayes, and PHYLIP. These can take genetic data (alignments) in the form of fastq files, fasta files, or NEXUS files.  


There are several different formats that a graphical tree can be built from.    
__Newick__: A collection of data formatted using specific syntax that includes parentheses, commas, and semicolons to delineate weight, time, and evolutionary distance. “Newick files are simply text files that consist of one or more tree descriptions in the Newick notation. In contrast to Nexus files they contain no further syntax elements or other information than the trees.”

### Comparing Trees

We will be comparing two graphical hypotheses of the genus _adelpha_, a group of butterflies.
![image](https://onlinelibrary.wiley.com/cms/asset/1d72eca6-ecb9-4b23-8230-868514c2fa01/mec13168-fig-0001-m.jpg)

One tree is morphologically based, from Keith Wilmott’s 2003 paper [“Cladistic analysis of the Neotropical butterfly genus Adelpha (Lepidoptera: Nymphalidae), with comments on the subtribal classification of Limenitidini.”](https://onlinelibrary.wiley.com/doi/abs/10.1046/j.1365-3113.2003.00209.x) Our molecular tree is from Emily Ebel’s 2015 paper [“Rapid diversification associated with ecological specialization in Neotropical Adelpha butterflies.”](https://onlinelibrary.wiley.com/doi/10.1111/mec.13168)

### Example

To start off, we want to understand how a tree is created. For this example we will be making our own tree using Newick format.

Here is a great explanation of Newick format: [__"Put simply, monophyletic clades are surrounded by parentheses and sister clades are separated by commas. For example, a simple tree could be written as (((A,B),C),(D,E))."__](http://ib.berkeley.edu/courses/ib200/2016/labs/02/lab02.pdf) Let's try making that!

You can check that you have the most up-to-date version of R by running the command "R.Version"
<div class = "blue">
```{r}
R.version
```
</div>
Once you know you have the correct version of R, install and load the following packages 
<div class = "blue">
```{r}
library(ape)
library(fastmatch)
library(quadprog)
library(phangorn)
library(phytools)
library(geiger)
```
</div>

You can make sure you have the most up-to-date version of each package by using the command "packageVersion("package-name")
<div class = "blue">
```{r}
packageVersion("ape")
```
</div>

Once we have all the necessary packages loaded into our markdown file, we can start playing around with building some phylogenetic trees! If you already know the relationships between the groups of species that you want to plot, and these the clade you are plotting isn't too complex, you can simply write out the tree as a text string in Newick format! Let's try it first with letters:  
<div class = "blue">
```{r set-up}
text.string<-
    "(((A,B),C),(D,E));"
example.tree<-read.tree(text=text.string)#this command reads trees in Newick format like we did above
plot(example.tree,no.margin=TRUE,edge.width=2) 
```  
</div>

Looks good! Now we can try it plotting a clade of whale species: 
<div class = "blue">
```{r animals}
text.string<-
    "((((humpback wahle, fin whale), (Antarctic minke whale, common minke whale)), bowhead whale), sperm whale);"
whale.tree<-read.tree(text=text.string)
plot(whale.tree,no.margin=TRUE,edge.width=2)
``` 
</div>

![We love to see it!](/Users/hamme/Documents/R/Super-Fly-Group-Module/image.png)

There are many different commands that will allow you to visualize your tree in many different ways. Let's try a few! 
<div class = "blue">
```{r}
roundPhylogram(whale.tree) #creates rounded branches in tree 
```
</div>
<div class = "blue">
```{r}
plot(unroot(whale.tree),type="unrooted",no.margin=TRUE,lab4ut="axial",
    edge.width=2) #creates an unrooted tree 
```
</div>
<div class = "blue">
```{r}
plotTree(whale.tree,type="fan",fsize=0.7,lwd=1,
    ftype="i") #creates a fan tree 
```
</div>

### Steps

1. <p class="text-primary">__Load the tree files into R__</p> 
  - Morphological data is in a NEXUS (.nex) file, and molecular data is in a .tre file in Newick format.
  
2. <p class="text-primary">__Adjust and manipulate tree using ggtree package to make graphics more readable__</p>

3. <p class="text-primary">__Comparing Trees__</p>  
  - Check to see if the trees are equal using all.equal(morphtree, moleculartreel) phytools package

### Molecular Tree

Now that we are a little more comfortable working with phylogenetic trees in R, we can load our first tree! 

First, Use this link ("https://github.com/sinnabunbun/Super-Fly-Group-Module") to go to the Super Fly Group Module repo. From there, click on the "NJst.tre" file. Copy this data using the little pencil icon. Then go to your own working repo, select the "Create New File" option and paste the tree data into that file.  

Once you have the tree file saved in your repo, you can load the tree into your R markdown file using the "read.nexus" command
<div class = "blue">
```{r}
mol.tree<-read.nexus(file="NJst.tre")
mol.tree
```
</div>

Once the tree is loaded, we can try plotting it! 
<div class = "blue">
```{r}
plotTree(mol.tree,ftype="i",fsize=0.6,lwd=2, no.margin = TRUE)
```
</div>

the Ntip() function will tell you how many different species (or tips) are represented in your tree 
<div class = "blue">
```{r}
Ntip(mol.tree) ##66 species in this tree
```
</div>

Just like with the whale tree above, we can use "unroot" to create an unrooted tree 
<div class = "blue">
```{r}
plot(unroot(mol.tree),type="unrooted",cex=0.6,
    use.edge.length=FALSE,lab4ut="axial",
    no.margin=TRUE)
##unrooted tree 
```
</div>

We can also make a fan tree 
<div class = "blue">
```{r}
plotTree(mol.tree,type="fan",fsize=0.7,lwd=2,
    ftype="i") 
```
</div>

If you want to see all the species names, use the code mol.tree$tip.label
<div class = "blue">
```{r}
##all the species names 
mol.tree$tip.label
```
</div>

You can also add arrows to draw attention to specific species! 
<div class = "blue">
```{r}
##add an arrow on a specific branch tip 
plotTree(mol.tree,type="fan",fsize=0.7,lwd=1,
    ftype="i")
add.arrow(mol.tree,tip="A_saundersii_saundersii",arrl=1)
```
</div>

### Morphological Tree

In order to compare molecular and morphological trees of the Adelpha lineage, we needed to build a tree in R from a phylogeny constructed based on morphological data, like the one reported in Keith Wilmott's paper. However, because this tree was created in 2003 and is therefore not in a format that is easily compatible with current R software, we ended up transcribing the tree into R in Newick format based on the phylogenetic relationships presented in Wilmott's Figure 8. 

<div class = "blue">
```{r}
string <- "((A. alala negra, A. corcyra aretina, A. tracta, A. pithys, A. donysa donysa), ((A. serpacelerio, A. seriphiaaquillia, A. seriphiatherasia), (A. melonaleucocoma, A. salmoneus colada, A. cythereacytherea, A. cythereadaguana, A. epioneagilla, A. etheldaethelda, A. thessaliathessalia, A. iphicleolathessalita, A. iphiclusiphiclus, A. shuara, A. plesaurephliassa, A. basiloides, A. atticaattica, A. leucerioides, A. saundersiisaundersii, A. boreasboreas, A. rothschildi, A. sichaeus, A. cocalacocala, A. leucophthalmairminella, A. irminatumida, A. pollina, A. lycoriaslara, A. lycoriasspruceana, A. erotia erotia, A. mesentina, A. phylaca pseudaethalia, A. capucinus capucinus, A. heraclea heraclea, A. naxia naxia, A. justina valentina, A. olynthia, A. jordani, A. zinazina, A. zinairma, A. delinita delinita, A. boeotia boeotia, A. maleaaethalia)));"

morph.tree<-read.tree(text=string)
plot(morph.tree,no.margin=TRUE,edge.width=2)
```
</div>

Just as we did with the molecular tree above, we can also plot this tree is a variety of ways 
<div class = "blue">
```{r}
morph.tree2 <- plot(morph.tree, "fan", main="Morphological Tree")
morph.tree2
```
</div>

or an unrooted tree 
<div class = "blue">
```{r}
morph.tree3 <- plot(morph.tree, "unrooted", main="Morphological Tree")
morph.tree3
```
</div>

## GGTree

`ggtree` is a complimentary package to `ggplot2` and will help us visualize our tree data. `ggtree` is no currently avaliable through CRAN so here is the code for installing the package from [BiocManager](https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html#fig:viewClade)
<div class = "blue">
```{r}
 if (!requireNamespace("BiocManager", quietly = TRUE))
+     install.packages("BiocManager")
 BiocManager::install("ggtree")
```
</div>

After installing, load `ggplot2` and `ggtree` into your markdown file. 
<div class = "blue">
```{r}
library(ggplot2)
library(ggtree)
```
</div>

There is a lot of great things this package has to offer and we would highly recommend looking at some of the commands that interface with ggplot! For example, the following code will help us change the color, size, and line type of our tree. We can also label the tips of each branch of the tree. Familiarize yourself with the commands, unfortunately there isn't an official `ggtree` cheat sheet... yet!
<div class = "blue">
```{r}
g <- ggtree(mol.tree, color="steelblue", size=0.5, linetype="dotted")
g <- g + geom_tiplab(size=2)
g
```
</div>

Similar to the different styles we used before to plot tree, `ggplot` alows for different types of rees as well.
<div class = "blue">
```{r}
ggtree(mol.tree, layout="circular") + ggtitle("(Phylogram) Circular Layout") + geom_tiplab(size=2)
```
</div>

We can also use `ggtree` to combine two trees for better visual comparison.
<div class = "blue">
```{r}
multiplot(ggtree(mol.tree) + geom_tiplab(size=2), ggtree(morph.tree) + geom_tiplab(size=2), ncol=2, labels=c('Molecular Tree', 'Morphological Tree'))
```
</div>

As you can see these trees are pretty large, the title is cut off in most views! Although in depth analysis might call for this many tip labels, we can also narrow down our species- which will help us narrow down what we're observing.

```{r}
reduced.species<-c("A.alalanegra", "A.atticaattica", "A.basiloides", "A.boeotiaboeotia", "A.boreasboreas", "A.capucinuscapucinus", "A.cocalacocala", "A.corcyraaretina", "A.cythereacytherea", "A.cythereadaguana", "A.delinitadelinita", "A.donysadonysa", "A.epioneagilla", "A.erotiaerotia", "A.heracleaheraclea", "A.iphicleolathessalita", "A.iphiclusiphiclus", "A.irminatumida", "A.jordani", "A.justinavalentina", "A.seriphiatherasia")
morph.reduced<-sapply(reduced.species,grep,morph.tree$tip.label)
morph.reduced
reduced.species<-morph.tree$tip.label[morph.reduced]
reduced.species
plotTree(morph.tree,type="fan",fsize=0.7,lwd=1,
    ftype="i")
add.arrow(morph.tree,tip=reduced.species,arrl=.5)
morph.tree.reduced<-drop.tip(morph.tree,
    setdiff(morph.tree$tip.label,reduced.species))
plotTree(morph.tree.reduced,ftype="i")

reduced.species.mol<-c("A_alala_negra", "A_attica_attica", "A_basiloides", "A_boeotia_boeotia", "A_boreas_boreas", "A_capucinus_capucinus", "A_cocala_cocala", "A_corcyra_aretina", "A_cytherea_cytherea", "A_cytherea_daguana", "A_delinita", "A_donysa_donysa", "A_epione_agilla", "A_erotia_erotia", "A_ethelda_ethelda", "A_heraclea", "A_iphicleola_thessalita", "A_iphiclus_iphiclus", "A_irmina_tumida", "A_jordani", "A_justina_valentina", "A_seriphia_therasia")
mol.reduced<-sapply(reduced.species.mol,grep,mol.tree$tip.label)
mol.reduced
reduced.species.mol<-mol.tree$tip.label[mol.reduced]
reduced.species.mol
plotTree(mol.tree,type="fan",fsize=0.7,lwd=1,
    ftype="i")
add.arrow(mol.tree,tip=reduced.species.mol,arrl=1)
mol.tree.reduced<-drop.tip(mol.tree,
    setdiff(mol.tree$tip.label,reduced.species.mol))
plotTree(mol.tree.reduced,ftype="i")
```

## Comparison

First we will use the function all.equal. This is a good place to start when comparing trees because it will tell you the basics: are these trees the same? It's a good exploratory function, but after we learn that they are not the same, we have to dig deeper.
<div class = "blue">
```{r}             
all.equal(mol.tree, morph.tree, use.edge.length = TRUE,
                   use.tip.label = TRUE)
```
</div>  
Next we'll try comparing phylogenetic tree topographies using the comparePhylo function. This will tell you that both trees are __NOT ultrametric__, which means the tips of each tree are not equally distant from the root of their respective tree. Essentially, both trees have structure and are not one big polytomy. It also tells us that they have different numbers of tips (46 and 66) which makes sense because they aren't looking at the same number of individuals. Naturally because of this, these trees will inherently be different from each other structurally. The morph tree has fewer nodes not only because it's hypothesizing the relationships among fewer individuals, but also because there is more uncertainty. This tree is comb like, and we can interpret this as an unresolved phylogeny.
<div class = "blue">
```{r}
comparePhylo(morph.tree, mol.tree, plot = FALSE, force.rooted = FALSE,
             use.edge.length = FALSE)
comparePhylo(morph.tree.reduced, mol.tree.reduced, plot = FALSE, force.rooted = FALSE, use.edge.length = FALSE)
```
</div>  
This serves to illustrate the importance of molecular data. Incomplete and outdated trees like our morphological example are very difficult to not only manipulate but to compare, which makes it hard to run statistical analysis on the differences between the two trees.